### PR TITLE
Homekit improvements

### PR DESCRIPTION
**Pull request by unknownskl**
I've made some improvements regarding the xboxone custom component.
These modifications will expose the following switches in Homekit:
- Power on/off
- Mute volume on/off

It also fixed to set the state of the xbox to on instead of unknown

### CODE REVIEW

**Comment by hunterjm** [comment]
The state is set to unknown because the “on” state doesn’t show media controls, and the Xbox doesn’t always report media as playing until you do a hard restart. In that case, HA 0.76 will show a “play/pause” button if the state is unknown.

File: xboxone/xboxone.py [hunk 1]
```diff
@@ -30,7 +30,7 @@
 SUPPORT_XBOXONE = SUPPORT_PAUSE | \
     SUPPORT_TURN_ON | SUPPORT_TURN_OFF | SUPPORT_PREVIOUS_TRACK | \
     SUPPORT_NEXT_TRACK | SUPPORT_SELECT_SOURCE | SUPPORT_PLAY | \
-    SUPPORT_VOLUME_STEP | SUPPORT_VOLUME_MUTE
+    SUPPORT_PAUSE | SUPPORT_VOLUME_STEP | SUPPORT_VOLUME_MUTE
```
Comment (hunterjm) [inline xboxone/xboxone.py]: SUPPORT_PAUSE is already defined on like 30.

File: xboxone/xboxone.py [hunk 2]
```diff
@@ -530,6 +537,10 @@ def turn_off(self):
 
     def mute_volume(self, mute):
         """Mute the volume."""
+        if self._muted == False:
```
Comment (hunterjm) [inline xboxone/xboxone.py]: This doesn’t work unless HA is the only thing able to mute the TV volume. The mute command is a toggle command, which means that if you were to mute your TV via your remote, your HA state would be incorrect.

Comment (tuxuser) [reply xboxone/xboxone.py]: Comparison to `True` or `False`should be done with `is`instead of `==`

**Comment by unknownskl** [comment]
Thanks for the feedback. I will dive deeper into the home assistant component to see if we can fix it on the long term. 
I will update the PR next weekend.

**Comment by dshokouhi** [comment]
@hunterjm I think we can make the states better.  Right now I think the usage of `STATE_UNKNOWN` is not helpful to the user.  This is what makes sense to me:

Unknown - device available but connection fails so user knows something is not working properly
On - if device is connected
Idle - when an app is loaded but nothing plays  (stopped is a good example)
Play/Pause - works as intended except blu-ray menu should really be considered Idle instead of playing but thats ok

These changes will bring things in line with Kodi and Plex media players.  Idle is an important status to go off if you ask me because how do you know the media stopped?

So line 454-456 should be STATE_IDLE which indicates a media ready app is loaded but nothing is playing or actively being paused.

Line 464 should be STATE_ON so user knows the xbox is on and connected.

**Comment by hunterjm** [comment]
I agree that those states make sense.  The only problem is with the front-end code [here](https://github.com/home-assistant/home-assistant-polymer/blob/master/src/util/hass-media-player-model.js#L38).

If you look at my front-end PR that was released in Home Assistant 0.76 (https://github.com/home-assistant/home-assistant-polymer/pull/1571) you will see that the Xbox does not always communicate media state over the smartglass protocol, and not all apps support the media state APIs, but still support play/pause.

Because of that, I defined the states as follows:
Off: No console available
Playing: Xbox returns playing in media_status
Paused: Xbox returns paused in media_status
Unknown: all other states

I then filter the available features based on the type of app that is loaded (here)[https://github.com/hunterjm/hassio-addons/blob/master/xboxone/xboxone.py#L442].

This allows the UI to still display a Play/Pause button when an app is loaded and attempt to send the play_pause_toggle command even if the app is not reporting a media status (which sometimes the console won't do until a hard restart).

@tuxuser - What are your thoughts?

**Comment by hunterjm** [comment]
We might be able to keep the state "ON" if no Application is loaded (i.e. you are playing a game), but even settings is considered an "Application".

**Comment by tuxuser** [comment]
Diplomatic answer: I am fine with everything that works good for everybody...
Didn't have the time to test different states intensively yet, just took a brief look so far.

**Comment by unknownskl** [comment]
I've processed the feedback.
For now i tried to find a good solution for the media state. Using STATE_ON when playing a game and STATE_PLAYING when using an app seems to be a reasonable way to do it.

**Comment by hunterjm** [comment]
I have been moving to a new house and haven’t had time to review. I’ll try to get to this ASAP.

**Comment by hunterjm** [comment]
I took a stab at this recently and think I found something that may work as a compromise.  It still uses the "Unknown" state for when there is no media state and the Xbox is in an "Application" because that is the true state of the "Media Player"... We don't know if media is playing or not, but the controls still work, so show a Play/Pause button (which only happens with UNKNOWN).

Look at the below example.  I was watching Netflix all day, and initially it was reporting the playing status back, but after a couple reboots it suddenly stopped reporting whether or not media was playing (it was).

image

img_2548

So here's what I'm going to go with:

1) Any media status other than Playing/Paused: **STATE_IDLE**
2) Media status of Playing: **STATE_PLAYING**
3) Media status of Paused: **STATE_PAUSED**
4) Console connected & available, but not in an Application: **STATE_ON**
5) Inside an Application with no media status: **STATE_UNKNOWN**
6) Console not available or connected: **STATE_OFF**

**Pull request closed without merging**